common/authorization/rsaTokenKeyProvider.go

```diff
@@ -0,0 +1,150 @@
+// The MIT License
+//
+// Copyright (c) 2020 Temporal Technologies Inc.  All rights reserved.
+//
+// Copyright (c) 2020 Uber Technologies, Inc.
+//
+// Permission is hereby granted, free of charge, to any person obtaining a copy
+// of this software and associated documentation files (the "Software"), to deal
+// in the Software without restriction, including without limitation the rights
+// to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+// copies of the Software, and to permit persons to whom the Software is
+// furnished to do so, subject to the following conditions:
+//
+// The above copyright notice and this permission notice shall be included in
+// all copies or substantial portions of the Software.
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+// THE SOFTWARE.
+
+package authorization
+
+import (
+	"crypto/ecdsa"
+	"crypto/rsa"
+	"encoding/json"
+	"fmt"
+	"net/http"
+	"strings"
+	"time"
+
+	"github.com/dgrijalva/jwt-go"
+
+	"go.temporal.io/server/common/log"
+	"go.temporal.io/server/common/log/loggerimpl"
+	"go.temporal.io/server/common/log/tag"
+	"go.temporal.io/server/common/service/config"
+)
+
+type jwks struct {
+	Keys []jsonWebKeys `json:"keys"`
+}
+
+type jsonWebKeys struct {
+	Kty string   `json:"kty"`
+	Kid string   `json:"kid"`
+	Use string   `json:"use"`
+	N   string   `json:"n"`
+	E   string   `json:"e"`
+	X5c []string `json:"x5c"`
+}
+
+// Default RSA token key provider
+type rsaKeyProvider struct {
+	config config.JWTKeyProvider
+	keys   map[string]*rsa.PublicKey
+	timer  *time.Timer
+	logger log.Logger
+}
+
+var _ TokenKeyProvider = (*rsaKeyProvider)(nil)
+
+func NewRSAKeyProvider(cfg *config.Config) *rsaKeyProvider {
+	logger := loggerimpl.NewLogger(cfg.Log.NewZapLogger())
+	provider := rsaKeyProvider{config: cfg.Global.Security.JWTKeyProvider, logger: logger}
+	provider.init()
+	return &provider
+}
+
+func (a *rsaKeyProvider) init() {
+	a.keys = make(map[string]*rsa.PublicKey)
+	if len(a.config.KeySourceURIs) == 0 {
```

Comment:
do you still want to start the timer in case an update configuration introduces new keys at a later point in time?

Reply:
I guess we don't have dynamic config...this is just going to update keys for existing URIs

Reply:
The ticker cost pretty much nothing perf wise. I'd err on the side of anticipating dynamic updates to config, especially for the hosted case. But I'm fine either way.